CHANGES/8540.bugfix.rst

```diff
@@ -0,0 +1,7 @@
+WSS: Fix heartbeat timeout logic -- by :user:`arcivanov`.
+
+The :py:meth:`~aiohttp.client_ws.ClientWebSocketResponse._pong_not_received` did not
+ terminate a :py:meth:`~aiohttp.client_ws.ClientWebSocketResponse.receive` operation.
```

Comment:
When I run the test, it seems like the .receive() terminates just fine. The only change is from receiving a WSCloseCode.OK to a WSCloseCode.ABNORMAL_CLOSURE.

Reply:
Well, firstly that's a huge deal - it's indication of nominal closure vs abnormal termination.
Secondly, I have a receive with a timeout, i.e. it will terminate naturally by itself.
Thirdly, it does terminate _in vivo_ as well during connection severance, just much later than the pong timeout is reached, where as feeding the `reader` an error message terminates a pending receive immediately as should the intention be.

Reply:
I have updated the test @Dreamsorcerer to simulate EOF not reaching either party (which is how hard-severed connection behaves) and the case is reproduced perfectly if you disable the `feed_data` stanza.

Reply:
OK, I think this looks about right now, but @bdraco is more familiar with the expected behaviour of websockets, so I'll wait for him to confirm.
We do need the master PR to go in first though.